### PR TITLE
Provide more information to the user

### PR DESCRIPTION
**Pull request by StarCycle**
1. remove abbreviations and use "longer names" in logs
2. enable progress bar for evaluation by default
3. enable asynchronized env by default
4. add readme intro for resuming training

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -251,6 +251,13 @@ checkpoints
 │   └── training_state.pth  # optimizer/scheduler/rng state and training step
 ```
 
+To resume training from a checkpoint, you can add these to the `train.py` python command:
+```bash
+    hydra.run.dir=your/original/experiment/dir resume=true
+```
+
+It will load the pretrained model, optimizer and scheduler states for training.
```
Comment (alexander-soare) [inline README.md:259]: How about adding: "For more information please see our tutorial on training resumption [here](./examples/5_resume_training.md)"

File: lerobot/configs/default.yaml [hunk 2]
```diff
@@ -120,7 +120,7 @@ eval:
   # `batch_size` specifies the number of environments to use in a gym.vector.VectorEnv.
   batch_size: 1
   # `use_async_envs` specifies whether to use asynchronous environments (multiprocessing).
-  use_async_envs: false
+  use_async_envs: true
```
Comment (alexander-soare) [inline lerobot/configs/default.yaml:123]: This doesn't work with the ACT environment, I think due to some EGL stuff. Which is why I defaulted it to `false`.

File: lerobot/scripts/eval.py [hunk 3]
```diff
@@ -210,8 +210,8 @@ def eval_policy(
     videos_dir: Path | None = None,
     return_episode_data: bool = False,
     start_seed: int | None = None,
-    enable_progbar: bool = False,
-    enable_inner_progbar: bool = False,
+    enable_progbar: bool = True,
```
Comment (alexander-soare) [inline lerobot/scripts/eval.py:213]: Can I check reasoning for changing these defaults? I believe you wanted to see the progbar during training, but in that case, why not pass the kwarg you want from `train.py`?

File: lerobot/scripts/train.py [hunk 4]
```diff
@@ -175,19 +175,19 @@ def log_train_info(logger: Logger, info, step, cfg, dataset, is_online):
     num_episodes = num_samples / avg_samples_per_ep
     num_epochs = num_samples / dataset.num_samples
     log_items = [
-        f"step:{format_big_number(step)}",
```
Comment (alexander-soare) [inline lerobot/scripts/train.py:178]: I think @Cadene will have an opinion about this. It's about trying to fit the output on one line. And I think I agree with him. I'm not sure the clarity is worth the ugliness of having this spill over onto two lines.

**Review by alexander-soare — CHANGES_REQUESTED** [review]
Thanks for these suggestions. I've left some comments.

File: README.md (hunk 1 above)
Comment (StarCycle) [reply README.md:259]: Nice idea!

File: lerobot/configs/default.yaml (hunk 2 above)
Comment (StarCycle) [reply lerobot/configs/default.yaml:123]: OK, I think we can let it to be `true` by default, by set it to `false` in the ACT env

File: lerobot/scripts/eval.py (hunk 3 above)
Comment (StarCycle) [reply lerobot/scripts/eval.py:213]: Yes I want to see the progbar during training. OK I will add them to `default.yaml` and pass the kwarg from `train.py`, that's reasonable

File: lerobot/scripts/train.py (hunk 4 above)
Comment (StarCycle) [reply lerobot/scripts/train.py:178]: Emmm @Cadene do you have spatial reasons to fit the output on one line? For example, you are developing a hardware with a very small screen, and you must fit all output within the screen?

Here is my story. I got stuck twice when I started training because:

1. The first time was the loooong data loading time because I didn't know what `data_s` meaned in the output log. If the log showed it was `dataloading time` more clearly, then I would know it was a dataloading problem and I could change `num_workers`, `batch_size`, etc. (I modified `batch_size` because I found the GPU utilization was too low
2. And then I found the looooong evaluation time because I didn't know I was using a sync env, instead of an asynv env. The logging just got stuck again...
3. Then I guessed perhaps I evaluated too many episodes. Thus, I set `eval.n_episodes=1` but I still got stuck again. The reason is this [line](https://github.com/huggingface/lerobot/blob/03d647269eff6d76df14c84159adc7624a582762/lerobot/scripts/eval.py#L241): 

```
n_batches = n_episodes // env.num_envs + int((n_episodes % env.num_envs) != 0)
# with n_episodes==1, we still get n_batches = 0 + 1 = 1
# Since we use 50 synchonized environments, we still evaluate 50 episodes one by one, even if the user sets n_episodes=1
```

If you do not need to fit the output in a small screen, I think the clarity is worth the ugliness:

### For a beginner user, he/she may only have 20 minutes in a weekday coffee break to try out LeRobot. If LeRobot is set up, run painlessly, and the initial output can be easily understood within the 20 minutes, the user will think it is a good tool and spends more time trying it later.

**Comment by StarCycle** [comment]
Perhaps you can also print the config at the beginning of training/eval.

Currently you are using hydra and different config files may override each other. Sometimes a user may not remember the setting in another config file, or not know his/her config is override by another config file. 

Simply printing all configs at the beginning of a training/eval process can solve this problem, like what they did for [mmengine](https://github.com/open-mmlab/mmengine).

File: lerobot/configs/default.yaml (hunk 2 above)
Comment (alexander-soare) [reply lerobot/configs/default.yaml:123]: Okay. You'll also have to set it to False in the Xarm env please.

File: lerobot/scripts/train.py (hunk 4 above)
Comment (alexander-soare) [reply lerobot/scripts/train.py:178]: Regarding your point 2, I think we can have some logic like:

```python
if eval.n_episodes < eval.batch_size:
    eval.batch_size = eval.n_episodes
```

Comment (StarCycle) [reply lerobot/scripts/train.py:178]: Nice idea, I just add it! I also set `async_env=False` for xarm env

File: lerobot/configs/default.yaml (hunk 2 above)
Comment (StarCycle) [reply lerobot/configs/default.yaml:123]: I just added it!

File: lerobot/configs/default.yaml [hunk 5]
```diff
@@ -120,7 +120,9 @@ eval:
...
+  episode_progbar: true
```
Comment (alexander-soare) [inline lerobot/configs/default.yaml:124]: Mind just putting these two in the train script directly hardcoded as function call params? I don't know that we want to bother with adding more config params for something so inconsequential.

File: lerobot/configs/env/aloha.yaml [hunk 6]
```diff
@@ -2,6 +2,12 @@
 
 fps: 50
 
+eval:
+  # `use_async_envs` specifies whether to use asynchronous environments (multiprocessing).
+  # set it to false to avoid EGL problem of the ACT env
```
Comment (alexander-soare) [inline lerobot/configs/env/aloha.yaml:7]: I'd rather not say this is an EGL problem without verifying that is truly the case. I'm not 100% sure it is. Maybe @aliberts knows?

File: lerobot/configs/env/dora_aloha_real.yaml [hunk 7]
```diff
@@ -2,6 +2,12 @@
 
 fps: 30
 
+eval:
```
Comment (alexander-soare) [inline lerobot/configs/env/dora_aloha_real.yaml:5]: I'm not sure this is relevant here as this is not simulation.

File: lerobot/scripts/train.py (hunk 4 above)
Comment (alexander-soare) [reply lerobot/scripts/train.py:178]: "Unresolving" this as we still have not heard from @Cadene.

I hear what you are saying but I still don't think it's worth it. IMO generally anyone who's advanced enough to consider timing bottlenecks, should be advanced enough to look into the code and see what these abbreviations mean. Most people who are not that advanced, wouldn't really notice it if it were written in plain english.

File: lerobot/scripts/eval.py [hunk 8]
```diff
@@ -453,6 +453,7 @@ def main(
         hydra_cfg = init_hydra_config(str(pretrained_policy_path / "config.yaml"), config_overrides)
     else:
         hydra_cfg = init_hydra_config(hydra_cfg_path, config_overrides)
+    logging.info(hydra_cfg)
```
Comment (alexander-soare) [inline lerobot/scripts/eval.py:456]: I'm not sure we want this on `eval.py`.

- It takes up a lot of lines on the terminal and when running multiple evals in a row (something I've done a lot in the past) you lose your place.
- Composition of multiple configs is not a problem here. You can just look at the config file in the pretrained model directory.
- Many of the params are relevant to training.

File: lerobot/scripts/train.py [hunk 9]
```diff
@@ -241,6 +241,7 @@ def train(cfg: DictConfig, out_dir: str | None = None, job_name: str | None = No
         raise NotImplementedError()
 
     init_logging()
+    logging.info(cfg)
```
Comment (alexander-soare) [inline lerobot/scripts/train.py:244]: This _could_ be useful, but right now it's just a massive one line json that's not very human readable. I think if we're going to do this it should be easier on the eyes.

File: lerobot/configs/env/aloha.yaml (hunk 6 above)
Comment (alexander-soare) [reply lerobot/configs/env/aloha.yaml:7]: I just tried running it and this is the error I see:

```
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":1"
      after 135 requests (135 known processed) with 0 events remaining.
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
Stepping through eval batches:   0%|                                                     | 0/1 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/home/alexander/Projects/lerobot/lerobot/scripts/eval.py", line 572, in <module>
    main(
  File "/home/alexander/Projects/lerobot/lerobot/scripts/eval.py", line 483, in main
    info = eval_policy(
  File "/home/alexander/Projects/lerobot/lerobot/scripts/eval.py", line 282, in eval_policy
    rollout_data = rollout(
  File "/home/alexander/Projects/lerobot/lerobot/scripts/eval.py", line 122, in rollout
    observation, info = env.reset(seed=seeds)
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/site-packages/gymnasium/vector/vector_env.py", line 140, in reset
    return self.reset_wait(seed=seed, options=options)
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/site-packages/gymnasium/vector/async_vector_env.py", line 249, in reset_wait
    results, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/site-packages/gymnasium/vector/async_vector_env.py", line 249, in <listcomp>
    results, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/multiprocessing/connection.py", line 250, in recv
    buf = self._recv_bytes()
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/multiprocessing/connection.py", line 414, in _recv_bytes
    buf = self._recv(4)
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/multiprocessing/connection.py", line 383, in _recv
    raise EOFError
EOFError
/home/alexander/anaconda3/envs/lerobot/lib/python3.10/site-packages/gymnasium/vector/async_vector_env.py:460: UserWarning: WARN: Calling `close` while waiting for a pending call to `reset` to complete.
  logger.warn(
Exception ignored in: <function AsyncVectorEnv.__del__ at 0x75b371771a20>
Traceback (most recent call last):
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/site-packages/gymnasium/vector/async_vector_env.py", line 549, in __del__
    self.close(terminate=True)
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/site-packages/gymnasium/vector/vector_env.py", line 272, in close
    self.close_extras(**kwargs)
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/site-packages/gymnasium/vector/async_vector_env.py", line 464, in close_extras
    function(timeout)
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/site-packages/gymnasium/vector/async_vector_env.py", line 249, in reset_wait
    results, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/site-packages/gymnasium/vector/async_vector_env.py", line 249, in <listcomp>
    results, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/multiprocessing/connection.py", line 250, in recv
    buf = self._recv_bytes()
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/multiprocessing/connection.py", line 414, in _recv_bytes
    buf = self._recv(4)
  File "/home/alexander/anaconda3/envs/lerobot/lib/python3.10/multiprocessing/connection.py", line 383, in _recv
    raise EOFError
EOFError:
```

maybe not EGL.

File: lerobot/scripts/train.py (hunk 4 above)
Comment (StarCycle) [reply lerobot/scripts/train.py:178]: You are right, to make it easier to understand, `number_of_seen_samples: 64` is better. If the general idea is approved, I will modify it.

I still hope it's as easy to understand as possible ^^

File: lerobot/configs/env/dora_aloha_real.yaml (hunk 7 above)
Comment (StarCycle) [reply lerobot/configs/env/dora_aloha_real.yaml:5]: Aha. My fault! will delete it

so dora is for real robotics only

File: lerobot/configs/env/aloha.yaml (hunk 6 above)
Comment (StarCycle) [reply lerobot/configs/env/aloha.yaml:7]: I see and let me modify it

File: lerobot/scripts/eval.py (hunk 8 above)
Comment (StarCycle) [reply lerobot/scripts/eval.py:456]: OK, let's remove it from eval.py

File: lerobot/scripts/train.py (hunk 9 above)
Comment (StarCycle) [reply lerobot/scripts/train.py:244]: A very good suggestion. Is there any way to print it in a human readable way? Can I still use `logging`? @alexander-soare

File: lerobot/scripts/train.py (hunk 4 above)
Comment (Cadene) [reply lerobot/scripts/train.py:178]: Instead of this change, I would prefer to update the [tutorial on `train.py`](https://github.com/huggingface/lerobot/blob/main/examples/4_train_policy_with_script.md), we should explain a typical output with some troubleshooting when data loading is high, etc.

We do this for the incoming `control_robot.py` tutorial:
Screenshot 2024-08-14 at 11 07 19

What do you think @alexander-soare ?

File: lerobot/configs/default.yaml (hunk 5 above)
Comment (Cadene) [reply lerobot/configs/default.yaml:124]: Out of curiosity, why do we want to make the progress bar an option?

File: README.md [hunk 10]
```diff
@@ -251,6 +251,13 @@ checkpoints
 │   └── training_state.pth  # optimizer/scheduler/rng state and training step
 ```
 
+To resume training from a checkpoint, you can add these to the `train.py` python command:
+```bash
+    hydra.run.dir=your/original/experiment/dir resume=true
+```
+
+It will load the pretrained model, optimizer and scheduler states for training. For more information please see our tutorial on training resumption [here](https://github.com/huggingface/lerobot/blob/main/examples/5_resume_training.md).
+
```
Comment (Cadene) [inline README.md:260]: Nice!

File: lerobot/common/utils/utils.py [hunk 11]
```diff
@@ -158,6 +158,13 @@ def init_hydra_config(config_path: str, overrides: list[str] | None = None) -> D
         version_base="1.2",
     )
     cfg = hydra.compose(Path(config_path).stem, overrides)
+    if cfg.eval.n_episodes < cfg.eval.batch_size:
+        print("init_hydra_config: eval.batch_size ({}) > eval.n_episodes ({}), set eval.batch_size to {}").format(
+            cfg.eval.batch_size,
+            cfg.eval.n_episodes,
+            cfg.eval.n_episodes,
+        )
+        cfg.eval.batch_size = cfg.eval.n_episodes
```
Comment (Cadene) [inline lerobot/common/utils/utils.py:167]: I am not sure to understand this code. Could we add a comment?

**Review by Cadene — COMMENTED** [review]
Thanks for your contribution! We will improve onboarding and ease of use.

Left a few comments

File: lerobot/common/utils/utils.py (hunk 11 above)
Comment (StarCycle) [reply lerobot/common/utils/utils.py:167]: Hi @Cadene,

As mentioned [here](https://github.com/huggingface/lerobot/pull/358#discussion_r1714645269), 
```
if cfg.eval.n_episodes < cfg.eval.batch_size:
    cfg.eval.batch_size = cfg.eval.n_episodes
```
is to avoid the program evaluates the policy for too many times, even if the user doesn't want to do so.

But you are right, I will add a comment here, instead of printing a message

**Comment by StarCycle** [comment]
@Cadene:

Re progressbar: you are right, I will not make it as an option. I still suggest to enable both progress bars (i.e., the progress bar for episodes and the bar for steps in an episode). Users can easily locate problems of evaluation if a step takes too long, or there are too many episodes

**Comment by StarCycle** [comment]
As a side note, now it logs this at the beginning of training, which is very easy to read:
```
INFO 2024-08-14 13:15:24       <stdin>:1 {'dataset_repo_id': 'lerobot/pusht',
 'device': 'cuda',
 'env': {'action_dim': 2,
         'episode_length': 300,
         'fps': '${fps}',
         'gym': {'obs_type': 'pixels_agent_pos',
                 'render_mode': 'rgb_array',
                 'visualization_height': 384,
                 'visualization_width': 384},
         'image_size': 96,
         'name': 'pusht',
         'state_dim': 2,
         'task': 'PushT-v0'},
 'eval': {'batch_size': 50, 'n_episodes': 50, 'use_async_envs': True},
 'fps': 10,
 'override_dataset_stats': {'action': {'max': [511.0, 511.0],
                                       'min': [12.0, 25.0]},
                            'observation.image': {'mean': [[[0.5]],
                                                           [[0.5]],
                                                           [[0.5]]],
                                                  'std': [[[0.5]],
                                                          [[0.5]],
                                                          [[0.5]]]},
                            'observation.state': {'max': [496.14618, 510.9579],
                                                  'min': [13.456424,
                                                          32.938293]}},
 'policy': {'beta_end': 0.02,
            'beta_schedule': 'squaredcos_cap_v2',
            'beta_start': 0.0001,
            'clip_sample': True,
            'clip_sample_range': 1.0,
            'crop_is_random': True,
            'crop_shape': [84, 84],
            'diffusion_step_embed_dim': 128,
            'do_mask_loss_for_padding': False,
            'down_dims': [512, 1024, 2048],
            'horizon': 16,
            'input_normalization_modes': {'observation.image': 'mean_std',
                                          'observation.state': 'min_max'},
            'input_shapes': {'observation.image': [3, 96, 96],
                             'observation.state': ['${env.state_dim}']},
            'kernel_size': 5,
            'n_action_steps': 8,
            'n_groups': 8,
            'n_obs_steps': 2,
            'name': 'diffusion',
            'noise_scheduler_type': 'DDPM',
            'num_inference_steps': None,
            'num_train_timesteps': 100,
            'output_normalization_modes': {'action': 'min_max'},
            'output_shapes': {'action': ['${env.action_dim}']},
            'prediction_type': 'epsilon',
            'pretrained_backbone_weights': None,
            'spatial_softmax_num_keypoints': 32,
            'use_film_scale_modulation': True,
            'use_group_norm': True,
            'vision_backbone': 'resnet18'},
 'resume': False,
 'seed': 100000,
 'training': {'adam_betas': [0.95, 0.999],
              'adam_eps': 1e-08,
              'adam_weight_decay': 1e-06,
              'batch_size': 64,
              'delta_timestamps': {'action': '[i / ${fps} for i in range(1 - '
                                             '${policy.n_obs_steps}, 1 - '
                                             '${policy.n_obs_steps} + '
                                             '${policy.horizon})]',
                                   'observation.image': '[i / ${fps} for i in '
                                                        'range(1 - '
                                                        '${policy.n_obs_steps}, '
                                                        '1)]',
                                   'observation.state': '[i / ${fps} for i in '
                                                        'range(1 - '
                                                        '${policy.n_obs_steps}, '
                                                        '1)]'},
              'do_online_rollout_async': False,
              'drop_n_last_frames': 7,
              'eval_freq': 100,
              'grad_clip_norm': 10,
              'image_transforms': {'brightness': {'min_max': [0.8, 1.2],
                                                  'weight': 1},
                                   'contrast': {'min_max': [0.8, 1.2],
                                                'weight': 1},
                                   'enable': False,
                                   'hue': {'min_max': [-0.05, 0.05],
                                           'weight': 1},
                                   'max_num_transforms': 3,
                                   'random_order': False,
                                   'saturation': {'min_max': [0.5, 1.5],
                                                  'weight': 1},
                                   'sharpness': {'min_max': [0.8, 1.2],
                                                 'weight': 1}},
              'log_freq': 200,
              'lr': 0.0001,
              'lr_scheduler': 'cosine',
              'lr_warmup_steps': 500,
              'num_workers': 4,
              'offline_steps': 200000,
              'online_buffer_capacity': None,
              'online_buffer_seed_size': 0,
              'online_env_seed': None,
              'online_rollout_batch_size': 1,
              'online_rollout_n_episodes': 1,
              'online_sampling_ratio': 0.5,
              'online_steps': 0,
              'online_steps_between_rollouts': 1,
              'save_checkpoint': True,
              'save_freq': 100},
 'use_amp': False,
 'video_backend': 'pyav',
 'wandb': {'disable_artifact': False,
           'enable': False,
           'notes': '',
           'project': 'lerobot'}}
```

**Comment by alexander-soare** [comment]
Thanks for revising this @StarCycle . My status is now "approving". I will also wait on @Cadene to approve as he has become involved.

Btw, looks like style tests are not passing. Have you seen CONTRIBUTING.md for instructions on how to set up the pre-commit hook?

**Review by alexander-soare — APPROVED** [review]

**Comment by Cadene** [comment]
@StarCycle By any chance, could you provide code to try this PR?

I feel like at least the third section is missing from the PR description among the sections we advise to add:
1. What the PR adds:
2. How it was tested
3. How to checkout & try? (for the reviewer) <--- example code

See this PR description for instance: https://github.com/huggingface/lerobot/pull/281

Thanks!

**Comment by StarCycle** [comment]
@Cadene 

You are right! I explain it here:

### What this does
1. Enable progress bar for evaluation by default, except in slurm
2. Enable asynchronized env by default, except for aloha environments.
3. Add readme intro for resuming training.
4. Add tutorial intro about explainations of abbreviations of the metrics in log
5. It will print all the configuration at the beginning of a training process

### How it was tested?
Not too much difference from the original code, just run `python lerobot/scripts/train.py policy=diffusion env=pusht` 

### How to checkout and try?
Just run `python lerobot/scripts/train.py policy=diffusion env=pusht`

File: lerobot/configs/env/aloha.yaml [hunk 12]
```diff
@@ -2,6 +2,12 @@
 
 fps: 50
 
+eval:
+  # `use_async_envs` specifies whether to use asynchronous environments (multiprocessing).
+  # set it to false to avoid some problems of the ACT env
+  # https://github.com/huggingface/lerobot/pull/358#discussion_r1716501659
+  use_async_envs: false
+
```
Comment (Cadene) [inline lerobot/configs/env/aloha.yaml:10]: See next comment on github link

**Review by Cadene — CHANGES_REQUESTED** [review]
I am very thankful for your time and suggestions.
I left some comments. I think some parts of this PR requires a bit more work.

On our side, we should do a better work at conveying the purpose for the two sections. They are actually quite important to make faster progress on LeRobot. I would have edited the PR description with these info:

## How it was tested?

- [x] Ran diffusion training and evaluation on pusht. Configs are displayed and look like this:
Screenshot 2024-08-16 at 16 06 09
Screenshot 2024-08-16 at 16 06 20

- [x] Ran diffusion training and evaluation on pusht. Progress bars are displayed and look like this:
Screenshot 2024-08-16 at 15 58 37

TODO:
- [ ] Run training and evaluation on aloha to validate `use_async_envs: false`
- [ ] Run training and evaluation on xarm to validate `use_async_envs: false`

## How to checkout and try?

- To checkout the code:
```bash
git remote add starcycle git@github.com:StarCycle/easier_lerobot.git
git fetch starcycle
git checkout starcycle/main
```

- To run diffusion training and evaluation on pusht:
```bash
python lerobot/scripts/train.py policy=diffusion env=pusht \
eval.batch_size=1 eval.use_async_envs=false training.eval_freq=2
```

- To run diffusion training and evaluation on aloha:
TODO

- To run diffusion training and evaluation on xarm:
TODO

Thanks for your help!!!!

File: lerobot/common/utils/utils.py [hunk 13]
```diff
@@ -158,7 +158,9 @@ def init_hydra_config(config_path: str, overrides: list[str] | None = None) -> D
         version_base="1.2",
     )
     cfg = hydra.compose(Path(config_path).stem, overrides)
-
+    if cfg.eval.n_episodes < cfg.eval.batch_size:
+        # Then we don't need some useless environments that comsume time and resources
+        cfg.eval.batch_size = cfg.eval.n_episodes
```
[the remark this answers is not in the document]
Comment (StarCycle) [reply lerobot/common/utils/utils.py]: I fully agree. This will make LeRobot easier to understand, which the PR is designed for

File: lerobot/configs/env/xarm.yaml [hunk 14]
```diff
@@ -2,6 +2,12 @@
 
 fps: 15
 
+eval:
+  # `use_async_envs` specifies whether to use asynchronous environments (multiprocessing).
+  # set it to false to avoid some problems of the ACT env
+  # https://github.com/huggingface/lerobot/pull/358#discussion_r1716501659
```
[the remark this answers is not in the document]
Comment (StarCycle) [reply lerobot/configs/env/xarm.yaml]: I see, I have fixed it

File: lerobot/scripts/train.py [hunk 15]
```diff
@@ -359,6 +360,8 @@ def evaluate_and_checkpoint_if_needed(step, is_online):
                     videos_dir=Path(out_dir) / "eval" / f"videos_step_{step_identifier}",
                     max_episodes_rendered=4,
                     start_seed=cfg.seed,
+                    enable_progbar=True,
+                    enable_inner_progbar=True,
```
[the remark this answers is not in the document]
Comment (StarCycle) [reply lerobot/scripts/train.py]: Aha...I just did it

Supporting slurm is a little surprising to me...does it support multi-gpu training at this moment?

**Comment by StarCycle** [comment]
Just out of curiosity, does LeRobot support multi-gpu training now? (you just mentioned slurm ^^

**Comment by Cadene** [comment]
@StarCycle Yes we are working on a PR using `accelerate`: https://github.com/huggingface/lerobot/pull/317

**Comment by StarCycle** [comment]
> @StarCycle Yes we are working on a PR using `accelerate`: #317

Nice!

**Comment by StarCycle** [comment]
Hi @Cadene,

Are there other things that I need to complete to merge this PR?

(ﾉ"◑ڡ◑)ﾉ

**Review by Cadene — APPROVED** [review]
Last update required and ready to merge!
